File: src/translator.zig
//! L2↔L3 Protocol Translator
//!
//! Handles bidirectional conversion between Layer 2 (Ethernet frames) and Layer 3 (IP packets).
//! Pure protocol implementation with no platform dependencies.

const std = @import("std");

/// Translator configuration
pub const TranslatorOptions = struct {
    our_mac: [6]u8,
    handle_arp: bool = true,
    learn_gateway_mac: bool = true,
    verbose: bool = false,
};

/// L2↔L3 Translator
pub const L2L3Translator = struct {
    allocator: std.mem.Allocator,
    options: TranslatorOptions,

    // Learned network information
    our_ip: ?u32, // Our IP address (learned from outgoing packets)
    gateway_ip: ?u32, // Gateway IP address
    gateway_mac: ?[6]u8, // Gateway MAC address (learned from packets)
    last_gateway_learn: i64, // Timestamp of last gateway MAC learn

    // Statistics
    packets_translated_l2_to_l3: u64,
    packets_translated_l3_to_l2: u64,
    arp_replies_learned: u64,

    const Self = @This();

    pub fn init(allocator: std.mem.Allocator, options: TranslatorOptions) !Self {
        return .{
            .allocator = allocator,
            .options = options,
            .our_ip = null,
            .gateway_ip = null,
            .gateway_mac = null,
            .last_gateway_learn = 0,
            .packets_translated_l2_to_l3 = 0,
            .packets_translated_l3_to_l2 = 0,
            .arp_replies_learned = 0,
        };
    }

    pub fn deinit(self: *Self) void {
        _ = self;
    }

    /// Convert IP packet (L3) to Ethernet frame (L2)
    /// Used when sending packets from TUN device to network/VPN that expects Ethernet frames
    ///
    /// **Memory Management:**
    /// This function allocates a new buffer for the Ethernet frame.
    /// The caller is responsible for freeing the returned slice using the same allocator.
    ///
    /// Returns: Allocated Ethernet frame (14-byte header + IP packet)
    /// Errors: InvalidPacket if the IP packet is malformed
    pub fn ipToEthernet(self: *Self, ip_packet: []const u8, dest_mac: ?[6]u8) ![]const u8 {
        if (ip_packet.len == 0) return error.InvalidPacket;

        // Determine EtherType and destination MAC
        var ethertype: u16 = undefined;
        var dst_mac: [6]u8 = undefined;

        if (ip_packet.len > 0 and (ip_packet[0] & 0xF0) == 0x40) {
            // IPv4 packet
            ethertype = 0x0800;

            // Learn our IP from source IP field (if not already known)
            if (self.our_ip == null and ip_packet.len >= 20) {
                const src_ip = std.mem.readInt(u32, ip_packet[12..16], .big);
                self.our_ip = src_ip;
                if (self.options.verbose) {
                    std.debug.print("[VirtualTap] Learned our IP: {}.{}.{}.{}\n", .{
                        (src_ip >> 24) & 0xFF,
                        (src_ip >> 16) & 0xFF,
                        (src_ip >> 8) & 0xFF,
                        src_ip & 0xFF,
                    });
                }
            }

            // Use provided dest MAC, otherwise gateway MAC, otherwise broadcast
            if (dest_mac) |dmac| {
                dst_mac = dmac;
            } else if (self.gateway_mac) |gw_mac| {
                dst_mac = gw_mac;
            } else {
                @memset(&dst_mac, 0xFF); // Broadcast
            }
        } else if (ip_packet.len > 0 and (ip_packet[0] & 0xF0) == 0x60) {
            // IPv6 packet
            ethertype = 0x86DD;
            if (dest_mac) |dmac| {
                dst_mac = dmac;
            } else {
                @memset(&dst_mac, 0xFF); // Broadcast for IPv6
            }
        } else {
            return error.InvalidPacket;
        }

        // Build Ethernet frame: [6 dest MAC][6 src MAC][2 EtherType][payload]
        const frame_size = 14 + ip_packet.len;
        const frame = try self.allocator.alloc(u8, frame_size);
        errdefer self.allocator.free(frame);

        @memcpy(frame[0..6], &dst_mac); // Destination MAC
        @memcpy(frame[6..12], &self.options.our_mac); // Source MAC
        std.mem.writeInt(u16, frame[12..14], ethertype, .big); // EtherType
        @memcpy(frame[14..], ip_packet); // IP packet

        self.packets_translated_l3_to_l2 += 1;

        return frame;
    }

    /// Convert Ethernet frame (L2) to IP packet (L3)
    /// Used when receiving Ethernet frames from network/VPN to write to TUN device
    ///
    /// **Memory Management:**
    /// - Returns `null` if the frame is non-IP (will be handled elsewhere, e.g., ARP)
    /// - Returns an **allocated** IP packet slice if conversion succeeded
    /// - The caller is responsible for freeing the returned slice using the same allocator
    ///
    /// Returns: Optional allocated IP packet slice (null if non-IP frame)
    /// Errors: InvalidPacket if the Ethernet frame is malformed
    pub fn ethernetToIp(self: *Self, eth_frame: []const u8) !?[]const u8 {
        if (eth_frame.len < 14) return error.InvalidPacket;

        const ethertype = std.mem.readInt(u16, eth_frame[12..14], .big);

        // Only handle IP packets, return null for ARP (caller handles it)
        if (ethertype != 0x0800 and ethertype != 0x86DD) {
            return null; // Not IP - will be handled by caller (ARP, etc.)
        }

        // Extract IP packet (strip 14-byte Ethernet header)
        const ip_packet = eth_frame[14..];

        // Learn gateway MAC from any packet from gateway IP
        if (ethertype == 0x0800 and ip_packet.len >= 20 and self.options.learn_gateway_mac) {
            const src_ip = std.mem.readInt(u32, ip_packet[12..16], .big);

            // If this packet is from our gateway, learn its MAC address
            if (self.gateway_ip) |gw_ip| {
                if (src_ip == gw_ip) {
                    var new_mac: [6]u8 = undefined;
                    @memcpy(&new_mac, eth_frame[6..12]); // Source MAC from Ethernet header

                    const changed = if (self.gateway_mac) |old_mac|
                        !std.mem.eql(u8, &old_mac, &new_mac)
                    else
                        true;

                    if (changed) {
                        self.gateway_mac = new_mac;
                        self.last_gateway_learn = std.time.milliTimestamp();
                        self.arp_replies_learned += 1;
                        if (self.options.verbose) {
                            std.debug.print("[VirtualTap] Learned gateway MAC: {X:0>2}:{X:0>2}:{X:0>2}:{X:0>2}:{X:0>2}:{X:0>2}\n", .{
                                new_mac[0], new_mac[1], new_mac[2], new_mac[3], new_mac[4], new_mac[5],
                            });
                        }
                    }
                }
            }
        }

        // Allocate copy of IP packet
        const result = try self.allocator.alloc(u8, ip_packet.len);
        @memcpy(result, ip_packet);

        self.packets_translated_l2_to_l3 += 1;

        return result;
    }

    /// Extract source IP from IP packet (for learning)
    pub fn extractSourceIp(ip_packet: []const u8) ?u32 {
        if (ip_packet.len < 20) return null;
        if ((ip_packet[0] & 0xF0) != 0x40) return null; // Not IPv4
        return std.mem.readInt(u32, ip_packet[12..16], .big);
    }

    /// Extract destination IP from IP packet
    pub fn extractDestIp(ip_packet: []const u8) ?u32 {
        if (ip_packet.len < 20) return null;
        if ((ip_packet[0] & 0xF0) != 0x40) return null; // Not IPv4
        return std.mem.readInt(u32, ip_packet[16..20], .big);
    }

    /// Check if IP packet is valid
    pub fn isValidIpPacket(packet: []const u8) bool {
        if (packet.len < 20) return false;
        const version = packet[0] >> 4;
        return version == 4 or version == 6;
    }

    /// Manually set our IP address
    pub fn setOurIp(self: *Self, ip: u32) void {
        self.our_ip = ip;
    }

    /// Manually set gateway IP
    pub fn setGatewayIp(self: *Self, gateway_ip: u32) void {
        self.gateway_ip = gateway_ip;
    }

    /// Manually set gateway MAC
    pub fn setGatewayMac(self: *Self, gateway_mac: [6]u8) void {
        self.gateway_mac = gateway_mac;
    }

    /// Get learned IP address
    pub fn getLearnedIp(self: *const Self) ?u32 {
        return self.our_ip;
    }

    /// Get learned gateway MAC
    pub fn getGatewayMac(self: *const Self) ?[6]u8 {
        return self.gateway_mac;
    }

    /// Get translation statistics
    pub fn getStats(self: *const Self) struct {
        l2_to_l3: u64,
        l3_to_l2: u64,
        arp_learned: u64,
    } {
        return .{
            .l2_to_l3 = self.packets_translated_l2_to_l3,
            .l3_to_l2 = self.packets_translated_l3_to_l2,
            .arp_learned = self.arp_replies_learned,
        };
    }
};

// Tests
test "L2L3Translator basic" {
    const allocator = std.testing.allocator;

    var translator = try L2L3Translator.init(allocator, .{
        .our_mac = [_]u8{ 0x00, 0x11, 0x22, 0x33, 0x44, 0x55 },
        .handle_arp = true,
        .learn_gateway_mac = true,
        .verbose = false,
    });
    defer translator.deinit();

    // Test IP to Ethernet
    const ip_packet = [_]u8{
        0x45, 0x00, 0x00, 0x54, // Version, IHL, TOS, Length
        0x00, 0x00, 0x40, 0x00, // ID, Flags
        0x40, 0x01, 0x00, 0x00, // TTL, Protocol, Checksum
        0xC0, 0xA8, 0x01, 0x64, // Source IP: 192.168.1.100
        0xC0, 0xA8, 0x01, 0x01, // Dest IP: 192.168.1.1
    };

    const eth_frame = try translator.ipToEthernet(&ip_packet, null);
    defer allocator.free(eth_frame);

    try std.testing.expectEqual(@as(usize, 14 + ip_packet.len), eth_frame.len);
    try std.testing.expectEqual(@as(u16, 0x0800), std.mem.readInt(u16, eth_frame[12..14], .big));
}

File: src/translator.c
#include "../include/virtual_tap_internal.h"

// ============================================================================
// Translator Implementation
// ============================================================================

Translator* translator_create(const uint8_t our_mac[6], bool handle_arp,
                              bool learn_gateway_mac, bool verbose) {
    if (!our_mac) return NULL;
    
    Translator* t = (Translator*)calloc(1, sizeof(Translator));
    if (!t) return NULL;
    
    memcpy(t->our_mac, our_mac, 6);
    t->our_ip = 0;
    t->gateway_ip = 0;
    memset(t->gateway_mac, 0, 6);
    memset(t->our_ipv6, 0, 16);
    memset(t->gateway_ipv6, 0, 16);
    t->has_ipv6 = false;
    t->has_ipv6_gateway = false;
    t->last_gateway_learn_ms = 0;
    t->handle_arp = handle_arp;
    t->learn_gateway_mac = learn_gateway_mac;
    t->verbose = verbose;
    t->packets_l2_to_l3 = 0;
    t->packets_l3_to_l2 = 0;
    t->arp_replies_learned = 0;
    
    return t;
}

void translator_destroy(Translator* t) {
    if (t) {
        free(t);
    }
}

// ============================================================================
// IP to Ethernet (L3 → L2)
// ============================================================================

int translator_ip_to_ethernet(Translator* t, const uint8_t* ip_packet, uint32_t ip_len,
                              const uint8_t* dest_mac, uint8_t* eth_out, uint32_t out_capacity) {
    if (!t || !ip_packet || !eth_out || ip_len == 0) {
        return VTAP_ERROR_INVALID_PARAMS;
    }
    
    if (out_capacity < ip_len + ETHERNET_HEADER_SIZE) {
        return VTAP_ERROR_BUFFER_TOO_SMALL;
    }
    
    // Detect IP version from first byte
    uint8_t version = (ip_packet[0] >> 4) & 0x0F;
    uint16_t ethertype;
    
    if (version == 4) {
        ethertype = ETHERTYPE_IPV4;
        
        // Learn our IP from source IP field (bytes 12-15)
        if (ip_len >= 20 && t->our_ip == 0) {
            t->our_ip = read_u32_be(ip_packet + 12);
            if (t->verbose) {
                printf("[Translator] Learned our IPv4 from outgoing packet\n");
            }
        }
    } else if (version == 6) {
        ethertype = ETHERTYPE_IPV6;
        
        // Learn our IPv6 from source address (bytes 8-23)
        if (ip_len >= 40 && !t->has_ipv6) {
            extract_ipv6_address(ip_packet, 8, t->our_ipv6);
            // Don't learn link-local addresses as primary
            if (!is_ipv6_link_local(t->our_ipv6)) {
                t->has_ipv6 = true;
                if (t->verbose) {
                    printf("[Translator] Learned our IPv6 from outgoing packet\n");
                }
            }
        }
    } else {
        return VTAP_ERROR_PARSE_FAILED;
    }
    
    // Determine destination MAC
    uint8_t dest[6];
    if (dest_mac) {
        memcpy(dest, dest_mac, 6);
    } else {
        // Check if gateway MAC is known
        bool has_gateway = false;
        for (int i = 0; i < 6; i++) {
            if (t->gateway_mac[i] != 0) {
                has_gateway = true;
                break;
            }
        }
        
        if (has_gateway) {
            memcpy(dest, t->gateway_mac, 6);
        } else {
            // Use broadcast
            memset(dest, 0xFF, 6);
        }
    }
    
    // Build Ethernet frame: [6 dest][6 src][2 type][IP payload]
    memcpy(eth_out, dest, 6);
    memcpy(eth_out + 6, t->our_mac, 6);
    write_u16_be(eth_out + 12, ethertype);
    memcpy(eth_out + ETHERNET_HEADER_SIZE, ip_packet, ip_len);
    
    t->packets_l3_to_l2++;
    
    return ip_len + ETHERNET_HEADER_SIZE;
}

// ============================================================================
// Ethernet to IP (L2 → L3)
// ============================================================================

int translator_ethernet_to_ip(Translator* t, const uint8_t* eth_frame, uint32_t eth_len,
                              uint8_t* ip_out, uint32_t out_capacity) {
    if (!t || !eth_frame || !ip_out || eth_len < ETHERNET_HEADER_SIZE) {
        return VTAP_ERROR_INVALID_PARAMS;
    }
    
    // Extract EtherType
    uint16_t ethertype = read_u16_be(eth_frame + 12);
    
    // Learn gateway MAC from source MAC if this is from gateway
    if (t->learn_gateway_mac && t->gateway_ip != 0) {
        // Check if source IP (for IPv4) matches gateway
        if (ethertype == ETHERTYPE_IPV4 && eth_len >= ETHERNET_HEADER_SIZE + 20) {
            uint32_t src_ip = read_u32_be(eth_frame + ETHERNET_HEADER_SIZE + 12);
            if (src_ip == t->gateway_ip) {
                const uint8_t* src_mac = eth_frame + 6;
                bool different = false;
                for (int i = 0; i < 6; i++) {
                    if (t->gateway_mac[i] != src_mac[i]) {
                        different = true;
                        break;
                    }
                }
                if (different) {
                    memcpy(t->gateway_mac, src_mac, 6);
                    t->last_gateway_learn_ms = get_time_ms();
                    if (t->verbose) {
                        printf("[Translator] Learned gateway MAC from incoming IPv4 packet\n");
                    }
                }
            }
        }
        
        // Check if source IPv6 matches gateway
        if (ethertype == ETHERTYPE_IPV6 && eth_len >= ETHERNET_HEADER_SIZE + 40 && t->has_ipv6_gateway) {
            uint8_t src_ipv6[16];
            extract_ipv6_address(eth_frame + ETHERNET_HEADER_SIZE, 8, src_ipv6);
            if (memcmp(src_ipv6, t->gateway_ipv6, 16) == 0) {
                const uint8_t* src_mac = eth_frame + 6;
                bool different = false;
                for (int i = 0; i < 6; i++) {
                    if (t->gateway_mac[i] != src_mac[i]) {
                        different = true;
                        break;
                    }
                }
                if (different) {
                    memcpy(t->gateway_mac, src_mac, 6);
                    t->last_gateway_learn_ms = get_time_ms();
                    if (t->verbose) {
                        printf("[Translator] Learned gateway MAC from incoming IPv6 packet\n");
                    }
                }
            }
        }
    }
    
    // Handle by EtherType
    if (ethertype == ETHERTYPE_IPV4 || ethertype == ETHERTYPE_IPV6) {
        uint32_t ip_len = eth_len - ETHERNET_HEADER_SIZE;
        
        if (out_capacity < ip_len) {
            return VTAP_ERROR_BUFFER_TOO_SMALL;
        }
        
        memcpy(ip_out, eth_frame + ETHERNET_HEADER_SIZE, ip_len);
        t->packets_l2_to_l3++;
        return ip_len;
    } else if (ethertype == ETHERTYPE_ARP) {
        // ARP handled separately
        return 0;
    } else {
        // Unknown protocol
        return 0;
    }
}

// ============================================================================
// Getters and Setters
// ============================================================================

uint32_t translator_get_our_ip(Translator* t) {
    return t ? t->our_ip : 0;
}

void translator_set_our_ip(Translator* t, uint32_t ip) {
    if (t) {
        t->our_ip = ip;
    }
}

uint32_t translator_get_gateway_ip(Translator* t) {
    return t ? t->gateway_ip : 0;
}

void translator_set_gateway_ip(Translator* t, uint32_t ip) {
    if (t) {
        t->gateway_ip = ip;
    }
}

bool translator_get_gateway_mac(Translator* t, uint8_t mac_out[6]) {
    if (!t || !mac_out) return false;
    
    bool has_mac = false;
    for (int i = 0; i < 6; i++) {
        if (t->gateway_mac[i] != 0) {
            has_mac = true;
            break;
        }
    }
    
    if (has_mac) {
        memcpy(mac_out, t->gateway_mac, 6);
    }
    
    return has_mac;
}

void translator_set_gateway_mac(Translator* t, const uint8_t mac[6]) {
    if (t && mac) {
        memcpy(t->gateway_mac, mac, 6);
        t->last_gateway_learn_ms = get_time_ms();
    }
}
